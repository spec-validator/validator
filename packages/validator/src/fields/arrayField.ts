import { Field, withErrorDecoration, declareField } from '../core'
import { Json } from '@spec-validator/utils/Json'

export interface ArrayField<T> extends Field<T[]> {
  readonly itemField: Field<T>,
}

export default declareField('@spec-validator/validator/fields/arrayField', <T>(itemField: Field<T>): ArrayField<T> => ({
  itemField,
  validate: (value: any): T[] => {
    if (!Array.isArray(value)) {
      throw 'Not an array'
    }
    return value.map(
      (it, index) => withErrorDecoration(index, () => itemField.validate(it))
    )
  },
  serialize: (deserialized: T[]): Json =>
    deserialized.map(
      (it, index) => withErrorDecoration(index, () => itemField.serialize(it) as unknown as Json)
    ),
}))
