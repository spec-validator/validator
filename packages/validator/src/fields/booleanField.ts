import { declareField } from '../core'
import { Json } from '@spec-validator/utils/Json'
import { FieldWithRegExpSupport } from './segmentField'

export type BooleanField = FieldWithRegExpSupport<boolean>

export default declareField('@spec-validator/validator/fields/booleanField', (): BooleanField => {
  const base = {
    validate: (value: any): boolean => {
      if (value !== true && value !== false) {
        throw 'Not a boolean'
      }
      return value
    },
    serialize: (deserialized: boolean): Json => deserialized,
    regex: /true|false|1|0/,
  }
  return {
    ...base,
    getStringField: () => ({
      ...base,
      validate: (value: any): boolean => {
        if (value === 'true' || value === '1') {
          return true
        }
        if (value === 'false' || value === '0') {
          return false
        }
        return base.validate(value)
      },
      serialize: (value: boolean): string => value.toString(),
    }),
  }

})
