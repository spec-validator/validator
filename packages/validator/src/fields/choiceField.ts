import { escapeRegex } from '@spec-validator/utils/utils'
import { Primitive, Json } from '@spec-validator/utils/Json'
import { declareField, OfType } from '../core'
import { FieldWithRegExp, FieldWithStringInputSupport } from './segmentField'

export interface ChoiceField<Choice extends Primitive> extends FieldWithStringInputSupport<Choice> {
  choices: readonly Choice[]
}

export default declareField('@spec-validator/fields.ChoiceField', <Choice extends Primitive>(
  ...choices: readonly Choice[]
): ChoiceField<Choice> => {
  const choicesSet = new Set(choices)

  const validate = (value: any): Choice => {
    if (choicesSet.has(value)) {
      return value as Choice
    }
    throw 'Invalid choice'
  }
  const serialize = (deserialized: Choice): Json => deserialized

  const result = {
    choices,
    validate,
    serialize,
  } as ChoiceField<Choice> & OfType<string>

  result.getFieldWithRegExp = ():
    Omit<ChoiceField<Choice>, 'getFieldWithRegExp'> & FieldWithRegExp<Choice> & OfType<string> => {

    const fullChoiceMap: Map<any, Primitive> = new Map<any, Primitive>()
    choices.forEach(it => {
      fullChoiceMap.set(it, it)
      fullChoiceMap.set(it.toString(), it)
    })

    return {
      type: result.type,
      choices,
      regex: new RegExp(
        Object.keys(choices)
          .map(it => it.toString())
          .map(escapeRegex)
          .join('|')
      ),
      serialize: (value: Choice) => value.toString(),
      validate: (value: any): Choice => validate(fullChoiceMap.get(value)),
    }
  }

  return result
})
