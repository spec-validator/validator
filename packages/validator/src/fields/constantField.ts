import { Primitive } from '@spec-validator/utils/Json'
import { escapeRegex } from '@spec-validator/utils/utils'
import { declareField } from '../core'
import { FieldWithRegExpSupport } from './segmentField'

export interface ConstantField<Constant extends Primitive> extends FieldWithRegExpSupport<Constant> {
  readonly constant: Constant
}

export default declareField('@spec-validator/validator/fields/constantField', <Constant extends Primitive> (
  constant: Constant
): ConstantField<Constant> => {
  const base = {
    regex: new RegExp(escapeRegex(constant.toString())),
    constant,
    validate: (value: any): Constant => {
      if (value !== constant) {
        throw 'Constant does not match the requirement'
      }
      return constant
    },
    serialize: () => constant,
  }

  return {
    ...base,
    getStringField: () => ({
      ...base,
      validate: (value: string): Constant => {
        if (value !== constant.toString()) {
          throw 'Constant does not match the requirement'
        }
        return constant
      },
      serialize: (value: Constant): string => value.toString(),
    }),
  }
})
