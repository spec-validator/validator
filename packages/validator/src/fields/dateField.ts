import { declareField } from '../core'

import { FieldWithRegExp, FieldWithStringInputSupport } from './segmentField'

const DateRegexps = {
  'date': /\d{4}-\d{2}-\d{2}/,
  'date-time': /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?Z/,
  'time': /\d{2}:\d{2}:\d{2}(\.\d{3})?Z/,
}

export interface DateField extends FieldWithStringInputSupport<Date>, FieldWithRegExp<Date> {
  format: keyof typeof DateRegexps
}

export default declareField('@spec-validator/fields.DateField', (
  format: keyof typeof DateRegexps = 'date-time'
): DateField => {
  const regex = DateRegexps[format]
  const result = {
    regex,
    validate: (value: any): Date => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      if (!regex.test(value)) {
        throw `Invalid ${format} string`
      }
      if (format === 'date-time') {
        return new Date(value)
      } else if (format === 'date') {
        return new Date(value)
      } else {
        return new Date('1970-01-01T' + value)
      }
    },
    serialize: (deserialized: Date) => {
      if (format === 'date-time') {
        return deserialized.toISOString()
      } else if (format === 'date') {
        return deserialized.toISOString().split('T')[0]
      } else {
        return deserialized.toISOString().split('T')[1]
      }
    },
  } as DateField

  result.getFieldWithRegExp = () => result
  result.format = format

  return result
})
