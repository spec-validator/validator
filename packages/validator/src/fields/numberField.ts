import { declareField, OfType } from '../core'
import { Json } from '@spec-validator/utils/Json'
import { FieldWithRegExp, FieldWithStringInputSupport } from './segmentField'

export interface NumberField extends FieldWithStringInputSupport<number> {
  canBeFloat: boolean
  signed: boolean
}

export default declareField('@spec-validator/fields.NumberField', ({ canBeFloat, signed }: {
  readonly canBeFloat?: boolean,
  readonly signed?: boolean
} = {}): NumberField => {
  const validate = (value: any): number => {
    if (typeof value !== 'number') {
      throw 'Not a number'
    }
    if (!canBeFloat && value !== Math.floor(value)) {
      throw 'Not an int'
    }
    if (!signed && value < 0) {
      throw 'Must be unsigned'
    }
    return value
  }
  const serialize = (deserialized: number): Json => deserialized

  const result = {
    canBeFloat,
    validate,
    serialize,
  } as NumberField & OfType<string>

  result.getFieldWithRegExp = ():
    Omit<NumberField, 'getFieldWithRegExp'> & FieldWithRegExp<number> & OfType<string> => {
    const parts: string[] = []
    if (signed) {
      parts.push('-?')
    }
    parts.push('\\d+')
    if (canBeFloat) {
      parts.push('(\\.\\d+)?')
    }

    return {
      canBeFloat: canBeFloat || false,
      signed: signed || false,
      type: result.type,
      validate: (value: any) => validate(Number.parseFloat(value)),
      serialize: (value: number) => value.toString(),
      regex: RegExp(parts.join('')),
    }
  }

  return result
})
