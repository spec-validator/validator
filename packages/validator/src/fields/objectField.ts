import { Field, TypeHint, withErrorDecoration, declareField } from '../core'
import { Json } from '@spec-validator/utils/Json'
import { Any } from '@spec-validator/utils/util-types'

export type ObjectFields<DeserializedType extends Record<string, Any> = Record<string, Any>> = {
  [P in keyof DeserializedType]: Field<DeserializedType[P]>
}

export interface ObjectField<
  Spec extends ObjectFields = ObjectFields
> extends Field<TypeHint<Spec>> {
  readonly objectSpec: Spec
  readonly canHaveExtraKeys: boolean
}

const ensureNoExtraFields = (
  validatorSpec: ObjectFields,
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  value: any
) => {
  const extraKeys = new Set(Object.keys(value))
  Object.keys(validatorSpec as any).forEach((it) => extraKeys.delete(it))
  if (extraKeys.size !== 0) {
    throw {
      extraKeys: Array.from(extraKeys),
    }
  }
}

export default declareField('@spec-validator/validator/fields/objectField', <
  Spec extends ObjectFields = ObjectFields
> (
    objectSpec: Spec,
    canHaveExtraKeys = false
  ): ObjectField<Spec> => ({
    objectSpec,
    canHaveExtraKeys,
    validate: (value: any): TypeHint<Spec> => {
      if (typeof value !== 'object' || value === null) {
        throw 'Not an object'
      }
      const result = Object.fromEntries(Object.entries(objectSpec).map(([key, valueSpec]) => [
        key, withErrorDecoration(key, () => valueSpec.validate(value[key])),
      ])) as TypeHint<Spec>
      if (!canHaveExtraKeys) {
        ensureNoExtraFields(objectSpec, value)
      }
      return result
    },
    serialize: (deserialized: TypeHint<Spec>): Json =>
    Object.fromEntries(Object.entries(objectSpec).map(([key, valueSpec]) => [
      key, withErrorDecoration(key, () => valueSpec.serialize(deserialized[key])),
    ])) as Json,
  }))

