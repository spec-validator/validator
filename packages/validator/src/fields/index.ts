import arrayField, { ArrayField } from './arrayField'
import booleanField, { BooleanField } from './booleanField'
import numberField, { NumberField } from './numberField'
import objectField, { ObjectField } from './objectField'
import stringField, { StringField } from './stringField'
import choiceField, { ChoiceField } from './choiceField'
import withDefault, { WithDefault } from './withDefault'
import optional, { OptionalField } from './optional'
import unionField, { UnionField } from './unionField'
import constantField, { ConstantField } from './constantField'
import segmentField, { SegmentField } from './segmentField'
import wildcardObjectField, { WildcardObjectField } from './wildcardObjectField'
import undefinedField, { UndefinedField } from './undefinedField'
import dateField, { DateField } from './dateField'

export {
  dateField, DateField,
  arrayField, ArrayField,
  booleanField, BooleanField,
  numberField, NumberField,
  objectField, ObjectField,
  stringField, StringField,
  choiceField, ChoiceField,
  withDefault, WithDefault,
  optional, OptionalField,
  unionField, UnionField,
  constantField, ConstantField,
  segmentField, SegmentField,
  wildcardObjectField, WildcardObjectField,
  undefinedField, UndefinedField,
}
