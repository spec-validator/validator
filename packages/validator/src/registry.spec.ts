import { Field } from '.'
import { numberField, stringField, objectField, booleanField } from './fields'
import createRegistry, { registryDeclaration as $, FieldPair } from './registry'

const normalRegistryPairs: FieldPair[] = [
  $(stringField, (field) => ({
    regexp: field.regex?.source,
    type: 'string',
  })),
  $(numberField, (field) => ({
    canBeFloat: field.params?.canBeFloat || false,
    type: 'number',
  })),
  $(objectField, (field, requestRepresentation) => Object.fromEntries(
    Object.entries(field.objectSpec).map(
      ([key, subfield]) => [key, requestRepresentation(subfield)]
    )
  )),
]

describe('createRegistry', () => {
  it('correctly returns a representation of a schema with known field', () => {
    const registry = createRegistry(normalRegistryPairs)
    const schema = objectField({
      title: stringField(),
      page: objectField({
        subtitle: stringField(),
        pageCount: numberField(),
      }),
    })
    expect(registry(schema)).toMatchSnapshot()
  })

  it('throws an error if a duplicate type is registered', () => {
    expect(() => createRegistry([...normalRegistryPairs, $(stringField, () => undefined)]))
      .toThrow(`Found duplicate of type declaration '${stringField.type}'`)
  })

  it('throws an error if field type is unknown', () => {
    const registry = createRegistry(normalRegistryPairs)
    const schema = objectField({
      flag: booleanField(),
    })
    expect(() => registry(schema))
      .toThrow(`Could not find field of type '${booleanField.type}'`)
  })

  it('throws an error if the field has no type', () => {
    const registry = createRegistry(normalRegistryPairs)
    const schema: Field<42> = {
      validate: () => 42,
      serialize: () => 42,
    }
    expect(() => registry(schema))
      .toThrow('Fields without \'type\' are not supported')
  })
})
