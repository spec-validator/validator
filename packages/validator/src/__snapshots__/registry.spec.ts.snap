// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRegistry correctly returns a representation of a schema with known field 1`] = `
Object {
  "page": Object {
    "pageCount": Object {
      "canBeFloat": false,
      "type": "number",
    },
    "subtitle": Object {
      "regexp": ".*",
      "type": "string",
    },
  },
  "title": Object {
    "regexp": ".*",
    "type": "string",
  },
}
`;
